cpp/src/arrow/util/bit_stream_utils.h

```diff
@@ -418,14 +450,65 @@ inline bool BitReader::GetVlqInt(uint32_t* v) {
 }
 
 inline bool BitWriter::PutZigZagVlqInt(int32_t v) {
-  auto u_v = ::arrow::util::SafeCopy<uint32_t>(v);
-  return PutVlqInt((u_v << 1) ^ (u_v >> 31));
+  uint32_t u_v = ::arrow::util::SafeCopy<uint32_t>(v);
+  v = (u_v << 1) ^ (v >> 31);
+  u_v = ::arrow::util::SafeCopy<uint32_t>(v);
```

Comment:
nit: just static_cast v >> 31 above.

Reply:
and elimnate this line.

Reply:
Actually, I'm not particularly clear about the difference between ::arrow::util::SafeCopy<uint32_t> and static_cast<uint32_t>. Could you please tell me when should we use SafeCopy rather than static_cast?

Reply:
Using safe-copy instead could also work.  SafeCopy (and other methods in ubsan.h) are  mostly to avoid undefined behavior.  The most common case of this is when trying to read possibly unaligned values.  

SafeCopy is equivelant to the new [std::bit_cast](https://en.cppreference.com/w/cpp/numeric/bit_cast)